approximate_string_matching/approximate_boyer_moore.py

```diff
@@ -0,0 +1,112 @@
+def boyer_moore_bad_environment(A, p, m, k):
+  bad_table = {(i, a): True for i in range(1, m + 1) for a in A}
+  for i in range(1, m + 1):
+    for j in range(max(0, i - k), min(i + k, m) + 1):
+      bad_table[(j, p[i])] = False
+  return bad_table
+
+
+def boyer_moore_multi_dim_shift(A, p, m, k):
+  ready = {a: m + 1 for a in A}
+  BM_k = {(i, a): m for a in A for i in range(m, m - k - 1, -1)}
+  for i in range(m - 1, 0, -1):
+    for j in range(ready[p[i]] - 1, max(i, m - k) - 1, -1):
+      BM_k[(j, p[i])] = j - i
+    ready[p[i]] = max(i, m - k)
+  return BM_k
+
+
+def approximate_boyer_moore(text, p, n, m, k=None, A=None):
+  if A is None:
+    A = set(list(text[1:] + p[1:]))
+
+  if k is None:
+    k = p.count('?')
```

Comment:
Please correct me if I'm wrong: if k is None and p contains '?', then does it solve matching with wildcards problem?

Reply:
No, unfortunately no. I just used it in test to determine default value of k. I followed the convention from 'dont_care' tests, thats why I put '?' in pattern. I thought widlcards could be easily referred to edit distance (deletion, swap, insertion).

Reply:
OK, let's then remove it and force user to provide k, so it won't confuse the reader with the wildcards (dont_care). I was definitely confused by that.

Reply:
Removed.